### ISSUE TITLE
Implementation does not match the Hao paper.

### ISSUE THREAD

**Issue by erlebach** [lucidrains/coconut-pytorch#6]
Your implementation creates thoughts that are trainable parameters and shared between samples. They are NOT generated from the output logits of the transformer as described in the paper. The thoughts are common to all samples. But in the Hao et al paper, thoughts are different for each sample. Could the authors of this codebase please explain further? The README appears to be incorrect in this regard.

**Comment by lucidrains** [comment]
Gordon, the thought tokens are simply the last layer hidden states fed back to the model recurrently, as done here https://github.com/lucidrains/coconut-pytorch/blob/main/coconut_pytorch/coconut.py#L396 could you point out where it is learned parameters shared across samples? you must be confusing the latent thought tokens with the begin of thought token.

**Comment by erlebach** [comment]
Thank you for responding! The following two lines: 
```
        self.begin_of_thought = nn.Parameter(torch.zeros(num_latents_per_step, dim))
        self.end_of_thought = nn.Parameter(torch.zeros(dim))
```
indicate that these two variables are trainable. You have `num_latents_per_step`, which is `c`  in the paper, which states: "As shown in Figure 2, in the initial stage, the model is trained on regular CoT instances. In the subsequent stages, at the k-th stage, the first k reasoning steps in the CoT are replaced with k × c continuous thoughts 1 , where c is a hyperparameter controlling the number of latent thoughts replacing a single language reasoning step."

The paper states: "Special tokens <bot> and <eot> are employed to mark the beginning and end of the latent thought mode, respectively.", which I interpret to mean by <bot> and <eot> are both of size `1 x d`. 

But I see what I missed (I think): 
```
def recurrent_latent_forward(
```
which has the comment: "        # latent reasoning is a recurrent model forward with the last hidden state being passed back in as input, while the prompt key / values are kept the same (prompt is NOT passed back in)"

I am sorry for missing this. Just one question; The algorithm does seem different from the paper though, in more minor ways (aside from your interesting use of hyperconnections, which I had not heard of). For example, the paper states that one starts with <bot><eot><step1><step2>...<answer> . Stage 1 leads to the context: <bot><thought1> <eot><step2><step3>...<answer>, and so  forth. So why would <bot> be multiple tokens? Perhaps I misunderstood the paper? 

Thanks again!

**Comment by lucidrains** [comment]
those are just two random ideas I threw in there, but can turned off and set to 1 respectively to be in line with the paper

**Comment by erlebach** [comment]
Great. Thank you. Random ideas are good, but perhaps a comment to that effect? :-) THank you for sharing.
